# Manipulating Data in Pandas

We won't cover every possible way to access data in this tutorial, but this should give you a sense of some of the main ways you can access and work with tabular data with Pandas. Things we'll cover:
- Selecting and accessing data from a DataFrame
- Filtering and reindexing data
- Transforming, sorting, aggregating, deduplicating

In [2]:
import pandas as pd
robocall_df = pd.read_csv("Data/Telemarketing_RoboCall_Weekly_Data_Transformed.csv")

We've loaded our data into a DataFrame which is much like a database table, or a single table in a spreadsheet. This table has rows and columns. Pandas has also added an index column which you'll see on the far left of the DataFrame. There is a LOT that a DataFrame can do - you can familiarize yourself with all it offers in the [documentation](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html). 

In [3]:
robocall_df

issues          time_issued  \
0      Telemarketing (including do not call and spoof...                  NaN   
1      Telemarketing (including do not call and spoof...  2014-07-01 08:20:00   
2      Telemarketing (including do not call and spoof...  2015-09-20 14:00:00   
3      Telemarketing (including do not call and spoof...  2015-09-23 11:46:00   
4                                              Robocalls  2015-09-28 15:19:00   
5      Telemarketing (including do not call and spoof...  2015-09-29 12:00:00   
6      Telemarketing (including do not call and spoof...  2015-10-01 11:00:00   
7      Telemarketing (including do not call and spoof...  2015-10-01 10:21:00   
8      Telemarketing (including do not call and spoof...  2015-10-01 13:02:00   
9      Telemarketing (including do not call and spoof...  2015-10-01 12:58:00   
10                                             Robocalls  2015-10-01 11:11:00   
11     Telemarketing (including do not call and spoof...  2015-10-01 13:30:00   
12     Telemarketing (including do not call and spoof...  2015-10-01 10:35:00   
13     Telemarketing (including do not call and spoof...  2015-10-01 06:50:00   
14     Telemarketing (including do not call and spoof...  2015-10-01 11:59:00   
15     Telemarketing (including do not call and spoof...  2015-10-01 10:49:00   
16                                             Robocalls  2015-10-01 10:51:00   
17     Telemarketing (including do not call and spoof...  2015-10-01 20:41:00   
18     Telemarketing (including do not call and spoof...  2015-10-01 11:21:00   
19     Telemarketing (including do not call and spoof...  2015-10-01 16:24:00   
20     Telemarketing (including do not call and spoof...  2015-09-25 11:30:00   
21                                             Robocalls                  NaN   
22                                             Robocalls  2015-10-01 12:00:00   
23                                             Robocalls  2015-10-01 13:10:00   
24     Telemarketing (including do not call and spoof...  2015-10-01 14:00:00   
25     Telemarketing (including do not call and spoof...  2015-10-01 13:45:00   
26     Telemarketing (including do not call and spoof...  2015-10-01 11:25:00   
27     Telemarketing (including do not call and spoof...  2015-10-01 14:58:00   
28     Telemarketing (including do not call and spoof...  2015-10-01 14:15:00   
29     Telemarketing (including do not call and spoof...  2015-09-28 09:43:00   
...                                                  ...                  ...   
38955  Telemarketing (including do not call and spoof...  2015-12-20 15:41:00   
38956                                          Robocalls  2015-12-19 12:23:00   
38957                                          Robocalls  2015-12-20 19:03:00   
38958  Telemarketing (including do not call and spoof...  2015-12-20 10:00:00   
38959  Telemarketing (including do not call and spoof...  2015-12-20 15:15:00   
38960                                          Robocalls  2015-12-20 16:54:00   
38961  Telemarketing (including do not call and spoof...  2015-12-15 08:35:00   
38962  Telemarketing (including do not call and spoof...  2015-12-20 20:00:00   
38963                                          Robocalls  2015-12-20 20:00:00   
38964                                          Robocalls  2015-12-20 09:15:00   
38965  Telemarketing (including do not call and spoof...  2015-12-20 20:59:00   
38966  Telemarketing (including do not call and spoof...  2015-12-14 14:46:00   
38967  Telemarketing (including do not call and spoof...  2015-12-14 15:34:00   
38968  Telemarketing (including do not call and spoof...  2015-12-16 14:32:00   
38969  Telemarketing (including do not call and spoof...  2015-12-16 17:26:00   
38970  Telemarketing (including do not call and spoof...  2015-12-16 18:51:00   
38971                                          Robocalls  2015-12-16 17:27:00   
38972  Telemarketing (including do not call and spoof...  2015-12-17 19:00:00   


If you just want a list of the column names that's easy enough:

In [4]:
robocall_df.columns

Index([u'issues', u'time_issued', u'caller_id', u'phone_number', u'type_robo',
       u'type_telemarketing', u'state', u'date_issued', u'date_created'],
      dtype='object')

To select a single column from the DataFrame you can use the name of the column within brackets:

In [5]:
robocall_df["phone_number"]

0                   -
1        610-990-4243
2                   -
3                   -
4                   -
5                   -
6                   -
7                   -
8                   -
9                   -
10                  -
11                  -
12                  -
13       321-608-4860
14                  -
15                  -
16                  -
17       847-234-9229
18                  -
19                  -
20                  -
21                  -
22                  -
23                  -
24       713-714-5831
25                  -
26       876-254-1483
27                  -
28       240-345-1622
29       786-275-5986
             ...     
38955               -
38956               -
38957               -
38958    201-380-0014
38959    815-264-2191
38960    800-203-0028
38961    312-637-5631
38962               -
38963               -
38964               -
38965               -
38966               -
38967               -
38968               -
38969     

You can get multiple columns by specifying them in a list.

In [6]:
robocall_df[["phone_number", "type_telemarketing"]]

phone_number type_telemarketing
0                 -                NaN
1      610-990-4243    Abandoned Calls
2                 -  Prerecorded Voice
3                 -    Abandoned Calls
4                 -                NaN
5                 -         Live Voice
6                 -         Live Voice
7                 -         Live Voice
8                 -  Prerecorded Voice
9                 -  Prerecorded Voice
10                -                NaN
11                -         Live Voice
12                -         Live Voice
13     321-608-4860    Abandoned Calls
14                -         Live Voice
15                -         Live Voice
16                -                NaN
17     847-234-9229         Live Voice
18                -         Live Voice
19                -  Prerecorded Voice
20                -  Prerecorded Voice
21                -                NaN
22                -                NaN
23                -                NaN
24     713-714-5831  Prerecorded Voice
25                -  Prerecorded Voice
26     876-254-1483         Live Voice
27                -  Prerecorded Voice
28     240-345-1622  Prerecorded Voice
29     786-275-5986         Live Voice
...             ...                ...
38955             -    Abandoned Calls
38956             -                NaN
38957             -                NaN
38958  201-380-0014         Live Voice
38959  815-264-2191         Live Voice
38960  800-203-0028                NaN
38961  312-637-5631         Live Voice
38962             -      Text Messaage
38963             -                NaN
38964             -                NaN
38965             -         Live Voice
38966             -    Abandoned Calls
38967             -    Abandoned Calls
38968             -    Abandoned Calls
38969             -    Abandoned Calls
38970             -    Abandoned Calls
38971  502-875-5130                NaN
38972             -    Abandoned Calls
38973             -    Abandoned Calls
38974             -    Abandoned Calls
38975             -    Abandoned Calls
38976             -    Abandoned Calls
38977             -    Abandoned Calls
38978             -    Abandoned Calls
38979             -         Live Voice
38980  209-813-4131                NaN
38981  772-446-0418                NaN
38982             -                NaN
38983             -    Abandoned Calls
38984             -         Live Voice

[38985 rows x 2 columns]

And if we want to get just one row of that column we can use a second set of brackets with the row index ("13" in the example below).

In [7]:
robocall_df["phone_number"][13]

'321-608-4860'

In some cases you might want to change a piece of data, for instance in the process of cleaning it up. So any edits you make directly to the dataframe will be reflected in the data. Verify in the output below that row index 13 has had it's phone_number updated. In other cases you may want to replace many values at once which can be done using the `.replace()` [function](http://pandas.pydata.org/pandas-docs/version/0.17.1/generated/pandas.DataFrame.replace.html)

In [8]:
robocall_df["phone_number"][13] = '404-608-4860'
robocall_df

issues          time_issued  \
0      Telemarketing (including do not call and spoof...                  NaN   
1      Telemarketing (including do not call and spoof...  2014-07-01 08:20:00   
2      Telemarketing (including do not call and spoof...  2015-09-20 14:00:00   
3      Telemarketing (including do not call and spoof...  2015-09-23 11:46:00   
4                                              Robocalls  2015-09-28 15:19:00   
5      Telemarketing (including do not call and spoof...  2015-09-29 12:00:00   
6      Telemarketing (including do not call and spoof...  2015-10-01 11:00:00   
7      Telemarketing (including do not call and spoof...  2015-10-01 10:21:00   
8      Telemarketing (including do not call and spoof...  2015-10-01 13:02:00   
9      Telemarketing (including do not call and spoof...  2015-10-01 12:58:00   
10                                             Robocalls  2015-10-01 11:11:00   
11     Telemarketing (including do not call and spoof...  2015-10-01 13:30:00   
12     Telemarketing (including do not call and spoof...  2015-10-01 10:35:00   
13     Telemarketing (including do not call and spoof...  2015-10-01 06:50:00   
14     Telemarketing (including do not call and spoof...  2015-10-01 11:59:00   
15     Telemarketing (including do not call and spoof...  2015-10-01 10:49:00   
16                                             Robocalls  2015-10-01 10:51:00   
17     Telemarketing (including do not call and spoof...  2015-10-01 20:41:00   
18     Telemarketing (including do not call and spoof...  2015-10-01 11:21:00   
19     Telemarketing (including do not call and spoof...  2015-10-01 16:24:00   
20     Telemarketing (including do not call and spoof...  2015-09-25 11:30:00   
21                                             Robocalls                  NaN   
22                                             Robocalls  2015-10-01 12:00:00   
23                                             Robocalls  2015-10-01 13:10:00   
24     Telemarketing (including do not call and spoof...  2015-10-01 14:00:00   
25     Telemarketing (including do not call and spoof...  2015-10-01 13:45:00   
26     Telemarketing (including do not call and spoof...  2015-10-01 11:25:00   
27     Telemarketing (including do not call and spoof...  2015-10-01 14:58:00   
28     Telemarketing (including do not call and spoof...  2015-10-01 14:15:00   
29     Telemarketing (including do not call and spoof...  2015-09-28 09:43:00   
...                                                  ...                  ...   
38955  Telemarketing (including do not call and spoof...  2015-12-20 15:41:00   
38956                                          Robocalls  2015-12-19 12:23:00   
38957                                          Robocalls  2015-12-20 19:03:00   
38958  Telemarketing (including do not call and spoof...  2015-12-20 10:00:00   
38959  Telemarketing (including do not call and spoof...  2015-12-20 15:15:00   
38960                                          Robocalls  2015-12-20 16:54:00   
38961  Telemarketing (including do not call and spoof...  2015-12-15 08:35:00   
38962  Telemarketing (including do not call and spoof...  2015-12-20 20:00:00   
38963                                          Robocalls  2015-12-20 20:00:00   
38964                                          Robocalls  2015-12-20 09:15:00   
38965  Telemarketing (including do not call and spoof...  2015-12-20 20:59:00   
38966  Telemarketing (including do not call and spoof...  2015-12-14 14:46:00   
38967  Telemarketing (including do not call and spoof...  2015-12-14 15:34:00   
38968  Telemarketing (including do not call and spoof...  2015-12-16 14:32:00   
38969  Telemarketing (including do not call and spoof...  2015-12-16 17:26:00   
38970  Telemarketing (including do not call and spoof...  2015-12-16 18:51:00   
38971                                          Robocalls  2015-12-16 17:27:00   
38972  Telemarketing (including do not call and spoof...  2015-12-17 19:00:00   


### Filtering & Reindexing

Let's go back to our original data by reloading the file.

In [9]:
robocall_df = pd.read_csv("Data/Telemarketing_RoboCall_Weekly_Data_Transformed.csv")

You notice that a lot of datasets you work with are deficient in some way or another. For instance, they may be missing values in some rows and columns. When it loads a file in Pandas is smart enough to mark empty fields as "NaN" which stands for Not a Number. 

In [10]:
robocall_df["type_telemarketing"]

0                      NaN
1          Abandoned Calls
2        Prerecorded Voice
3          Abandoned Calls
4                      NaN
5               Live Voice
6               Live Voice
7               Live Voice
8        Prerecorded Voice
9        Prerecorded Voice
10                     NaN
11              Live Voice
12              Live Voice
13         Abandoned Calls
14              Live Voice
15              Live Voice
16                     NaN
17              Live Voice
18              Live Voice
19       Prerecorded Voice
20       Prerecorded Voice
21                     NaN
22                     NaN
23                     NaN
24       Prerecorded Voice
25       Prerecorded Voice
26              Live Voice
27       Prerecorded Voice
28       Prerecorded Voice
29              Live Voice
               ...        
38955      Abandoned Calls
38956                  NaN
38957                  NaN
38958           Live Voice
38959           Live Voice
38960                  NaN
3

We can test for these values using the ``isnull`` and ``notnull`` functions which will return a True / False value based on the value of the item. 

In [11]:
robocall_df["type_telemarketing"].isnull()

0         True
1        False
2        False
3        False
4         True
5        False
6        False
7        False
8        False
9        False
10        True
11       False
12       False
13       False
14       False
15       False
16        True
17       False
18       False
19       False
20       False
21        True
22        True
23        True
24       False
25       False
26       False
27       False
28       False
29       False
         ...  
38955    False
38956     True
38957     True
38958    False
38959    False
38960     True
38961    False
38962    False
38963     True
38964     True
38965    False
38966    False
38967    False
38968    False
38969    False
38970    False
38971     True
38972    False
38973    False
38974    False
38975    False
38976    False
38977    False
38978    False
38979    False
38980     True
38981     True
38982     True
38983    False
38984    False
Name: type_telemarketing, dtype: bool

And we may want to filter out those empty values. We can do that with a special selector syntax. In the following notice that within the brackets we tell it to select rows for which type_telemarketing is not null. Another useful function for removing missing data is `dropna()` which has [parameters](http://pandas.pydata.org/pandas-docs/version/0.17.1/generated/pandas.DataFrame.dropna.html) that allow you to drop rows or columns have have any or all values that are missing. 

In [12]:
robocall_df[robocall_df["type_telemarketing"].notnull()]

issues          time_issued  \
1      Telemarketing (including do not call and spoof...  2014-07-01 08:20:00   
2      Telemarketing (including do not call and spoof...  2015-09-20 14:00:00   
3      Telemarketing (including do not call and spoof...  2015-09-23 11:46:00   
5      Telemarketing (including do not call and spoof...  2015-09-29 12:00:00   
6      Telemarketing (including do not call and spoof...  2015-10-01 11:00:00   
7      Telemarketing (including do not call and spoof...  2015-10-01 10:21:00   
8      Telemarketing (including do not call and spoof...  2015-10-01 13:02:00   
9      Telemarketing (including do not call and spoof...  2015-10-01 12:58:00   
11     Telemarketing (including do not call and spoof...  2015-10-01 13:30:00   
12     Telemarketing (including do not call and spoof...  2015-10-01 10:35:00   
13     Telemarketing (including do not call and spoof...  2015-10-01 06:50:00   
14     Telemarketing (including do not call and spoof...  2015-10-01 11:59:00   
15     Telemarketing (including do not call and spoof...  2015-10-01 10:49:00   
17     Telemarketing (including do not call and spoof...  2015-10-01 20:41:00   
18     Telemarketing (including do not call and spoof...  2015-10-01 11:21:00   
19     Telemarketing (including do not call and spoof...  2015-10-01 16:24:00   
20     Telemarketing (including do not call and spoof...  2015-09-25 11:30:00   
24     Telemarketing (including do not call and spoof...  2015-10-01 14:00:00   
25     Telemarketing (including do not call and spoof...  2015-10-01 13:45:00   
26     Telemarketing (including do not call and spoof...  2015-10-01 11:25:00   
27     Telemarketing (including do not call and spoof...  2015-10-01 14:58:00   
28     Telemarketing (including do not call and spoof...  2015-10-01 14:15:00   
29     Telemarketing (including do not call and spoof...  2015-09-28 09:43:00   
31     Telemarketing (including do not call and spoof...  2015-10-01 17:00:00   
32     Telemarketing (including do not call and spoof...  2015-10-01 11:03:00   
33     Telemarketing (including do not call and spoof...  2015-10-01 16:15:00   
34     Telemarketing (including do not call and spoof...  2015-10-01 15:05:00   
37     Telemarketing (including do not call and spoof...  2015-09-10 09:00:00   
38     Telemarketing (including do not call and spoof...  2015-09-27 18:53:00   
41     Telemarketing (including do not call and spoof...  2015-09-30 16:53:00   
...                                                  ...                  ...   
38938  Telemarketing (including do not call and spoof...  2015-12-18 17:29:00   
38939  Telemarketing (including do not call and spoof...  2015-12-18 14:17:00   
38940  Telemarketing (including do not call and spoof...  2015-12-20 08:15:00   
38944  Telemarketing (including do not call and spoof...  2015-12-15 09:10:00   
38945  Telemarketing (including do not call and spoof...  2015-12-18 10:28:00   
38948  Telemarketing (including do not call and spoof...  2015-12-19 16:11:00   
38949  Telemarketing (including do not call and spoof...  2015-12-20 10:16:00   
38950  Telemarketing (including do not call and spoof...  2015-12-20 20:20:00   
38953  Telemarketing (including do not call and spoof...  2015-12-18 15:22:00   
38955  Telemarketing (including do not call and spoof...  2015-12-20 15:41:00   
38958  Telemarketing (including do not call and spoof...  2015-12-20 10:00:00   
38959  Telemarketing (including do not call and spoof...  2015-12-20 15:15:00   
38961  Telemarketing (including do not call and spoof...  2015-12-15 08:35:00   
38962  Telemarketing (including do not call and spoof...  2015-12-20 20:00:00   
38965  Telemarketing (including do not call and spoof...  2015-12-20 20:59:00   
38966  Telemarketing (including do not call and spoof...  2015-12-14 14:46:00   
38967  Telemarketing (including do not call and spoof...  2015-12-14 15:34:00   
38968  Telemarketing (including do not call and spoof...  2015-12-16 14:32:00   


If you need to filter by more than one column you can combine them using the ``&`` character. Though note you need an extra pair of parentheses around each logical test. Let's grab this dataframe and assign it to another variable. 

In [13]:
maryland_df = robocall_df[(robocall_df["type_telemarketing"].notnull()) & (robocall_df["state"]=="Maryland")]
maryland_df

issues          time_issued  \
617    Telemarketing (including do not call and spoof...  2015-08-27 11:40:00   
618    Telemarketing (including do not call and spoof...  2015-09-25 13:24:00   
619    Telemarketing (including do not call and spoof...  2015-09-29 15:39:00   
620    Telemarketing (including do not call and spoof...  2015-09-30 10:40:00   
622    Telemarketing (including do not call and spoof...  2015-09-30 09:56:00   
623    Telemarketing (including do not call and spoof...  2015-09-30 17:01:00   
624    Telemarketing (including do not call and spoof...  2015-10-01 15:00:00   
625    Telemarketing (including do not call and spoof...  2015-10-01 14:14:00   
626    Telemarketing (including do not call and spoof...  2015-10-01 17:30:00   
629    Telemarketing (including do not call and spoof...  2015-10-01 09:50:00   
631    Telemarketing (including do not call and spoof...  2015-10-01 12:20:00   
632    Telemarketing (including do not call and spoof...  2015-10-01 14:06:00   
633    Telemarketing (including do not call and spoof...  2015-10-01 17:14:00   
636    Telemarketing (including do not call and spoof...  2015-10-01 17:34:00   
1225   Telemarketing (including do not call and spoof...  2015-02-26 18:32:00   
1226   Telemarketing (including do not call and spoof...  2015-08-19 18:35:00   
1227   Telemarketing (including do not call and spoof...  2015-08-20 09:52:00   
1229   Telemarketing (including do not call and spoof...  2015-08-20 09:46:00   
1230   Telemarketing (including do not call and spoof...  2015-09-15 14:02:00   
1232   Telemarketing (including do not call and spoof...  2015-09-30 09:00:00   
1233   Telemarketing (including do not call and spoof...  2015-09-30 08:45:00   
1234   Telemarketing (including do not call and spoof...  2015-09-30 10:24:00   
1235   Telemarketing (including do not call and spoof...  2015-10-01 09:03:00   
1240   Telemarketing (including do not call and spoof...  2015-10-02 13:39:00   
1241   Telemarketing (including do not call and spoof...  2015-10-02 22:00:00   
1242   Telemarketing (including do not call and spoof...  2015-10-02 09:58:00   
1243   Telemarketing (including do not call and spoof...  2015-10-02 13:30:00   
1244   Telemarketing (including do not call and spoof...  2015-10-02 14:34:00   
1473   Telemarketing (including do not call and spoof...  2015-09-26 15:44:00   
1474   Telemarketing (including do not call and spoof...  2015-10-02 15:00:00   
...                                                  ...                  ...   
37550  Telemarketing (including do not call and spoof...  2015-12-15 20:17:00   
37551  Telemarketing (including do not call and spoof...  2015-12-16 09:00:00   
37552  Telemarketing (including do not call and spoof...  2015-12-16 12:00:00   
37553  Telemarketing (including do not call and spoof...  2015-12-16 11:50:00   
37554  Telemarketing (including do not call and spoof...  2015-12-16 12:56:00   
37555  Telemarketing (including do not call and spoof...  2015-12-16 20:00:00   
37561  Telemarketing (including do not call and spoof...  2015-12-16 18:13:00   
38160  Telemarketing (including do not call and spoof...  2015-12-16 10:03:00   
38162  Telemarketing (including do not call and spoof...  2015-12-17 10:30:00   
38163  Telemarketing (including do not call and spoof...  2015-12-17 15:30:00   
38164  Telemarketing (including do not call and spoof...  2015-12-17 16:40:00   
38165  Telemarketing (including do not call and spoof...  2015-12-17 16:25:00   
38167  Telemarketing (including do not call and spoof...  2015-12-17 14:04:00   
38169  Telemarketing (including do not call and spoof...  2015-12-17 18:45:00   
38170  Telemarketing (including do not call and spoof...  2015-12-17 17:37:00   
38171  Telemarketing (including do not call and spoof...  2015-12-17 09:24:00   
38679  Telemarketing (including do not call and spoof...  2015-12-11 12:40:00   
38680  Telemarketing (including do not call and spoof...  2015-12-16 11:00:00   


After all that filtering you might wonder how much data you have left. To check the shape (i.e. number of rows and columns) of a DataFrame just append ``.shape`` at the end. 

In [14]:
maryland_df.shape

(774, 9)

You'll notice that in the filtered data frame the index starts from "617", but maybe we want to reset it to start at zero now that we're focused on Maryland. We can do that, but remember we have to assign the new dataframe back to the same name (i.e. `maryland_df`)

In [15]:
maryland_df = maryland_df.reset_index(drop=True)
maryland_df

issues          time_issued  \
0    Telemarketing (including do not call and spoof...  2015-08-27 11:40:00   
1    Telemarketing (including do not call and spoof...  2015-09-25 13:24:00   
2    Telemarketing (including do not call and spoof...  2015-09-29 15:39:00   
3    Telemarketing (including do not call and spoof...  2015-09-30 10:40:00   
4    Telemarketing (including do not call and spoof...  2015-09-30 09:56:00   
5    Telemarketing (including do not call and spoof...  2015-09-30 17:01:00   
6    Telemarketing (including do not call and spoof...  2015-10-01 15:00:00   
7    Telemarketing (including do not call and spoof...  2015-10-01 14:14:00   
8    Telemarketing (including do not call and spoof...  2015-10-01 17:30:00   
9    Telemarketing (including do not call and spoof...  2015-10-01 09:50:00   
10   Telemarketing (including do not call and spoof...  2015-10-01 12:20:00   
11   Telemarketing (including do not call and spoof...  2015-10-01 14:06:00   
12   Telemarketing (including do not call and spoof...  2015-10-01 17:14:00   
13   Telemarketing (including do not call and spoof...  2015-10-01 17:34:00   
14   Telemarketing (including do not call and spoof...  2015-02-26 18:32:00   
15   Telemarketing (including do not call and spoof...  2015-08-19 18:35:00   
16   Telemarketing (including do not call and spoof...  2015-08-20 09:52:00   
17   Telemarketing (including do not call and spoof...  2015-08-20 09:46:00   
18   Telemarketing (including do not call and spoof...  2015-09-15 14:02:00   
19   Telemarketing (including do not call and spoof...  2015-09-30 09:00:00   
20   Telemarketing (including do not call and spoof...  2015-09-30 08:45:00   
21   Telemarketing (including do not call and spoof...  2015-09-30 10:24:00   
22   Telemarketing (including do not call and spoof...  2015-10-01 09:03:00   
23   Telemarketing (including do not call and spoof...  2015-10-02 13:39:00   
24   Telemarketing (including do not call and spoof...  2015-10-02 22:00:00   
25   Telemarketing (including do not call and spoof...  2015-10-02 09:58:00   
26   Telemarketing (including do not call and spoof...  2015-10-02 13:30:00   
27   Telemarketing (including do not call and spoof...  2015-10-02 14:34:00   
28   Telemarketing (including do not call and spoof...  2015-09-26 15:44:00   
29   Telemarketing (including do not call and spoof...  2015-10-02 15:00:00   
..                                                 ...                  ...   
744  Telemarketing (including do not call and spoof...  2015-12-15 20:17:00   
745  Telemarketing (including do not call and spoof...  2015-12-16 09:00:00   
746  Telemarketing (including do not call and spoof...  2015-12-16 12:00:00   
747  Telemarketing (including do not call and spoof...  2015-12-16 11:50:00   
748  Telemarketing (including do not call and spoof...  2015-12-16 12:56:00   
749  Telemarketing (including do not call and spoof...  2015-12-16 20:00:00   
750  Telemarketing (including do not call and spoof...  2015-12-16 18:13:00   
751  Telemarketing (including do not call and spoof...  2015-12-16 10:03:00   
752  Telemarketing (including do not call and spoof...  2015-12-17 10:30:00   
753  Telemarketing (including do not call and spoof...  2015-12-17 15:30:00   
754  Telemarketing (including do not call and spoof...  2015-12-17 16:40:00   
755  Telemarketing (including do not call and spoof...  2015-12-17 16:25:00   
756  Telemarketing (including do not call and spoof...  2015-12-17 14:04:00   
757  Telemarketing (including do not call and spoof...  2015-12-17 18:45:00   
758  Telemarketing (including do not call and spoof...  2015-12-17 17:37:00   
759  Telemarketing (including do not call and spoof...  2015-12-17 09:24:00   
760  Telemarketing (including do not call and spoof...  2015-12-11 12:40:00   
761  Telemarketing (including do not call and spoof...  2015-12-16 11:00:00   
762  Telemarketing (including do not call and spoof...  2015-12-18 10:00:00   
763  Telemarketing 

### Accessing Rows

We may also sometimes need to access a row of data from a data frame. This can be done with the `iloc` accessor and providing the integer-based position in brackets, or with the `loc` accessesor and providing the label-based index in brackets. In this example the two are equivalent.

In [27]:
maryland_df.iloc[0]

issues                Telemarketing (including do not call and spoof...
time_issued                                         2015-08-27 11:40:00
caller_id                                                             -
phone_number                                               202-370-6665
type_robo                                                           NaN
type_telemarketing                                           live voice
state                                                          Maryland
date_issued                                                  2015-08-27
date_created                                                 2015-10-01
Name: 0, dtype: object

And if we need that row as an array we can use the `.values` which is sometimes necessary is we want to do other types of mathematical operations:

In [24]:
maryland_df.iloc[0].values

array(['Telemarketing (including do not call and spoofing)',
       '2015-08-27 11:40:00', '-', '202-370-6665', nan, 'live voice',
       'Maryland', '2015-08-27', '2015-10-01'], dtype=object)

### Applying Data Transformations

Sometimes you will want to transform your data by applying a transformation function to each datum within a column or row. We don't necessarily need to, but to show you how to do it, let's make all the text in the `type_telemarketing` column lowercase. We define a function which takes in an input datum (x in this case) and returns the transformed value of that. We use the `apply` [function](http://pandas.pydata.org/pandas-docs/version/0.17.1/generated/pandas.DataFrame.apply.html) on the dataframe to apply that function to an entire column (or to an entire row).

In [18]:
def lowercaser(x):
    return x.lower()

maryland_df["type_telemarketing"] = maryland_df["type_telemarketing"].apply(lowercaser)

### Sorting 

Oftentimes you will want to sort your data to get an overview or see what is at the top or bottom of a ranking. To sort by values use the `sort_values` [function](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html#pandas.DataFrame.sort_values). Below we sort by the time_issued column from most recent to least recent. 

In [19]:
maryland_df.sort_values(by="time_issued", ascending=False)

issues          time_issued  \
771  Telemarketing (including do not call and spoof...  2015-12-19 21:00:00   
772  Telemarketing (including do not call and spoof...  2015-12-19 18:51:00   
773  Telemarketing (including do not call and spoof...  2015-12-19 12:48:00   
770  Telemarketing (including do not call and spoof...  2015-12-19 09:08:00   
763  Telemarketing (including do not call and spoof...  2015-12-18 16:45:00   
765  Telemarketing (including do not call and spoof...  2015-12-18 16:00:00   
769  Telemarketing (including do not call and spoof...  2015-12-18 15:25:00   
764  Telemarketing (including do not call and spoof...  2015-12-18 14:50:00   
766  Telemarketing (including do not call and spoof...  2015-12-18 13:15:00   
767  Telemarketing (including do not call and spoof...  2015-12-18 13:07:00   
768  Telemarketing (including do not call and spoof...  2015-12-18 12:45:00   
762  Telemarketing (including do not call and spoof...  2015-12-18 10:00:00   
757  Telemarketing (including do not call and spoof...  2015-12-17 18:45:00   
758  Telemarketing (including do not call and spoof...  2015-12-17 17:37:00   
754  Telemarketing (including do not call and spoof...  2015-12-17 16:40:00   
755  Telemarketing (including do not call and spoof...  2015-12-17 16:25:00   
753  Telemarketing (including do not call and spoof...  2015-12-17 15:30:00   
756  Telemarketing (including do not call and spoof...  2015-12-17 14:04:00   
752  Telemarketing (including do not call and spoof...  2015-12-17 10:30:00   
759  Telemarketing (including do not call and spoof...  2015-12-17 09:24:00   
749  Telemarketing (including do not call and spoof...  2015-12-16 20:00:00   
750  Telemarketing (including do not call and spoof...  2015-12-16 18:13:00   
748  Telemarketing (including do not call and spoof...  2015-12-16 12:56:00   
746  Telemarketing (including do not call and spoof...  2015-12-16 12:00:00   
747  Telemarketing (including do not call and spoof...  2015-12-16 11:50:00   
761  Telemarketing (including do not call and spoof...  2015-12-16 11:00:00   
751  Telemarketing (including do not call and spoof...  2015-12-16 10:03:00   
745  Telemarketing (including do not call and spoof...  2015-12-16 09:00:00   
744  Telemarketing (including do not call and spoof...  2015-12-15 20:17:00   
741  Telemarketing (including do not call and spoof...  2015-12-15 12:26:00   
..                                                 ...                  ...   
656  Telemarketing (including do not call and spoof...  2015-08-01 08:00:00   
513  Telemarketing (including do not call and spoof...  2015-06-01 10:25:00   
14   Telemarketing (including do not call and spoof...  2015-02-26 18:32:00   
735  Telemarketing (including do not call and spoof...  2014-09-12 18:40:00   
94   Telemarketing (including do not call and spoof...  2014-01-01 10:00:00   
83   Telemarketing (including do not call and spoof...                  NaN   
127  Telemarketing (including do not call and spoof...                  NaN   
142  Telemarketing (including do not call and spoof...                  NaN   
161  Telemarketing (including do not call and spoof...                  NaN   
164  Telemarketing (including do not call and spoof...                  NaN   
172  Telemarketing (including do not call and spoof...                  NaN   
187  Telemarketing (including do not call and spoof...                  NaN   
191  Telemarketing (including do not call and spoof...                  NaN   
281  Telemarketing (including do not call and spoof...                  NaN   
282  Telemarketing (including do not call and spoof...                  NaN   
297  Telemarketing (including do not call and spoof...                  NaN   
298  Telemarketing (including do not call and spoof...                  NaN   
311  Telemarketing (including do not call and spoof...                  NaN   
346  Telemarketing (including do not call and spoof...                  NaN   
355  Telemarketing 

### Aggregation 

You'll often want to summarize DataFrames to get an overview of your data, or to aggregate it. The `describe()` function is useful for an initial overview, but there are many others such as `min()`, `max()`, `sum()`, `mean()`, and [many others](http://pandas.pydata.org/pandas-docs/stable/basics.html#descriptive-statistics)

In [20]:
maryland_df.describe()

issues  \
count                                                 774   
unique                                                  1   
top     Telemarketing (including do not call and spoof...   
freq                                                  774   

                time_issued caller_id phone_number          type_robo  \
count                   749       774          774                  4   
unique                  737       494          138                  2   
top     2015-11-10 12:50:00         -            -  Prerecorded Voice   
freq                      2       166          634                  3   

       type_telemarketing     state date_issued date_created  
count                 774       774         774          774  
unique                  5         1          95           76  
top            live voice  Maryland  2015-11-16   2015-10-06  
freq                  295       774          22           22

A useful analytic operation is to create groups that can then be summarized. This can be accomplished with the `groupby()` [function](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.groupby.html). Various [aggregation functions](http://pandas.pydata.org/pandas-docs/stable/groupby.html) can then be applied. 

In [21]:
state_groups = robocall_df.groupby("state")
state_groups.count()

issues  time_issued  caller_id  phone_number  type_robo  \
state                                                                           
Alabama                  505          481        505           505        134   
Alaska                    40           38         40            40         10   
American Samoa             1            1          1             1          1   
Arizona                  890          850        890           890        334   
Arkansas                 245          237        245           245         69   
California              5246         5019       5246          5246       1719   
Colorado                1030          985       1030          1030        359   
Connecticut              672          645        672           672        234   
Delaware                 141          133        141           141         34   
District of Columbia     202          192        202           202         60   
Florida                 2851         2728       2851          2851        912   
Georgia                 1321         1261       1321          1321        388   
Guam                       3            3          3             3          1   
Hawaii                    76           76         76            76         19   
Idaho                    169          164        169           169         48   
Illinois                1423         1361       1423          1423        436   
Indiana                  584          554        584           584        213   
Iowa                     313          297        313           313        129   
Kansas                   311          298        311           311         96   
Kentucky                 515          497        515           515        112   
Louisiana                348          337        348           348         87   
Maine                    105           98        105           105         39   
Maryland                1091         1042       1091          1091        317   
Massachusetts           1069         1012       1069          1069        376   
Michigan                 989          949        989           989        328   
Minnesota                663          638        663           663        235   
Mississippi              226          215        226           226         59   
Missouri                 477          465        477           477        144   
Montana                  110          107        110           110         35   
Nebraska                 237          228        237           237         97   
Nevada                   361          342        361           361        105   
New Hampshire            190          186        190           190         60   
New Jersey              1106         1057       1106          1106        355   
New Mexico               273          259        273           273         76   
New York                2181         2070       2181          2181        637   
North Carolina          1078         1019       1078          1078        358   
North Dakota              74           72         74            74         28   
Ohio                    1108         1068       1108          1108        330   
Oklahoma                 498          482        498           498        197   
Oregon                   585          560        585           585        192   
Pennsylvania            1605         1529       1605          1605        529   
Puerto Rico               16           14         16            16          6   
Rhode Island              94           90         94            94         29   
South Carolina           493          468        493           493        174   
South Dakota              58           54         58            58         20   
Tennessee                861          818        861           861        259   
Texas                   2669         2563       2669          2669        735   
Utah                     391          371        391 

### Deduplication

At times your data will have duplicate rows in it and you'll want to remove those. To check for duplicated rows you can use the `.duplicated()` [function](http://pandas.pydata.org/pandas-docs/version/0.17.1/generated/pandas.DataFrame.duplicated.html), and if you want to check for duplicates within a certain column you can pass that as a parameter. Let's say we want to detect duplicate caller id numbers: 

In [22]:
maryland_df.duplicated(["caller_id"])

0      False
1      False
2      False
3      False
4      False
5      False
6      False
7      False
8      False
9      False
10      True
11     False
12     False
13     False
14     False
15     False
16     False
17      True
18     False
19     False
20     False
21     False
22      True
23     False
24      True
25     False
26     False
27     False
28      True
29     False
       ...  
744     True
745     True
746    False
747    False
748    False
749    False
750    False
751    False
752    False
753     True
754    False
755    False
756     True
757    False
758    False
759    False
760     True
761    False
762     True
763    False
764     True
765    False
766    False
767    False
768     True
769     True
770    False
771    False
772    False
773     True
dtype: bool

We can then drop the rows detected as duplicates using the `drop_duplicates()` [function](http://pandas.pydata.org/pandas-docs/version/0.17.1/generated/pandas.DataFrame.drop_duplicates.html). 

In [23]:
maryland_df.drop_duplicates(["caller_id"]).shape

(494, 9)